### ISSUE TITLE
Incorrect/inconsistent equality of custom units

### ISSUE THREAD

**Issue by randyheydon** [astropy/astropy#12609]
### Description
```
>>> import astropy.units as u
>>> Hz = u.def_unit('Hz', u.cycle / u.s)
>>> kHz = u.def_unit('kHz', 1000 * Hz)
>>> 1 * kHz == 1000 * Hz
True
>>> 1000 * Hz == 1 * kHz
False
```
Of course, both equalities should return True, and in general equalities should be commutative.

(As an aside, I understand why Astropy defines Hz as 1/s, but in my field of work, cycle/s is more convenient and less error-prone, so I want to be able to redefine it.)

### System Details
This was found on an Anaconda install on Windows:
- Windows-10-10.0.19042-SP0
- Numpy 1.20.3
- pyerfa 2.0.0
- astropy 5.0
- Scipy 1.7.1
- Matplotlib 3.5.0

**Comment by mhvk** [comment]
That looks puzzling indeed! But it looks like the underlying reason is a rounding error in converting from the right-hand unit to the left-hand unit (which must happen for the equality test):
```
1000*Hz-1*kHz
# <Quantity 1.13686838e-13 Hz>
kHz.to(Hz)
# 999.9999999999999
```
This looks even more confusing at first, but can be understood by looking at the code at https://github.com/astropy/astropy/blob/9ef8d4f2d9f36139ff84e76ce6f500e38f4dc257/astropy/units/core.py#L1068-L1096

Here, the decomposition is the trick: cycle to radian is what brings about the rounding problem.

Now looking at this again, I slightly wonder why this was set up like this; I'll see if dividing the units before decomposing is better and faster...

**Comment by mhvk** [comment]
p.s. Agreed that often cy/s is much handier; see #6433 about a long-standing issue about this. It would still be nice if we somehow could give the user a choice... (and then wavelengths should probably also be in m/cycle)

**Comment by mhvk** [comment]
Since the main problem is round-off error, I removed the "bug" label.

**Comment by hugobuddel** [comment]
It is a bit annoying though, e.g. `um.to(nm)` is also `999.9999999999999`, and `mm.to(um)` is `1000.0000000000001`.

Maybe there is a smarter way for handling these standard SI prefixes.

**Comment by mhvk** [comment]
It is a pity, indeed, but the basic problem is not in astropy:
```
In [19]: 1e-6/1e-9
Out[19]: 999.9999999999999
```

**Comment by hugobuddel** [comment]
Depends on the perspective. It could be considered within scope of astropy to deal with the unit conversion accurately, and that the root problem is that astropy relies on float division.

That is, arguably, astropy should do the conversion exactly whenever possible, and only rely on floating point operations when necessary. However, this code has been there for more than a decade, and apparently this concern is not shared by many others, so maybe the code is fine as it is.

There would also be speed concerns to changing this. I'll write my own slower-but-more-accurate unit conversion; if I do I'll share it here.

**Comment by randyheydon** [comment]
@hugobuddel the issue may not simply be float division, but rather that astropy is always using conversions to the same base set of units.  E.g. in your um-to-nm example, it uses conversions to meters; in my Hz-to-kHz example, it uses conversions to radians.  That means that, as units get farther from the base, they necessarily lose accuracy in conversion.  And it shows that a simple (but slow) fix like switching from floats to rationals isn't sufficient to maintain accuracy, because some conversions are irrational (like cycles-to-radians).

Floats could still be sufficiently accurate _if_ a shortest-path conversion could be done instead.  This means keeping track of how each unit is defined relative to others and walking through that graph, rather than going all the way to a base unit and back.

I'd love to see your take on unit conversion code if you get a chance to complete it!

**Comment by mhvk** [comment]
Indeed, part of of it is going by the decomposed unit, though also:
```
In [12]: u.m.to(u.nm)
Out[12]: 999999999.9999999
```
which is a consequence of
```
In [48]: 1/1e-9
Out[48]: 999999999.9999999
```
In the end, I think indeed that the conclusion on our side has been that we just live with it.  Things are accurate as long as you do not print all digits, as is automatically done when dealing with arrays:
```
In [49]: (1*u.m).to(u.nm)
Out[49]: <Quantity 1.e+09 nm>
```
Also, in practice quantities are usually not nearly-integer in some unit to start with, and almost all calculations involving them have uncertainties, so it doesn't really matter.

**Comment by hugobuddel** [comment]
If the quantities have integer values (or float values that represent integer value), then it would be nice to have exact conversions though, whenever possible.

But I agree; we should keep the code as is, because it would require more thought to do this right than just hacking the `_to()` function.  E.g. it might be worthwhile to try to also getting the Hz thing right.

Nevertheless, I thought of several ways to accomplish exact conversion, for my immediate goal of mm <-> um <-> nm. The (enormous) kludge that I settled on, makes use of the fact that SI unit conversions have short human readable representations by design. I'm only showing it here because I promised, and just maybe it could help someone:

```python
def save_unit_to(unit_from: u.Unit, unit_to: u.Unit):
    """Like unit_from.to(unit_to), but u.um to u.nm == 1000., not 999.999.."""
    to1 = unit_from.to(unit_to)
    sto1 = str(to1)
    to2 = 1 / unit_to.to(unit_from)
    sto2 = str(to2)
    if len(sto2) < len(sto1):
        return to2
    return to1
```

with as test:
```python
def test_save_to():
    for ufrom, uto, fac in [
        (u.um, u.nm, 1000.),
        (u.mm, u.um, 1000.),
        (u.nm, u.um, 0.001),
        (u.um, u.mm, 0.001),
    ]:
        assert save_unit_to(ufrom, uto) == fac
```

Maybe someone will say "Noooooooo......." and implement it properly.

**Comment by eerovaher** [comment]
Like @mhvk already mentioned in practice the floating point errors are smaller than measurement uncertainties, so they don't affect the correctness of computations. When presenting the value of a `Quantity` to a human then `1000` is indeed preferable to `999.9999999999999`, but the correct tool for this job is [formatting](https://docs.astropy.org/en/stable/units/format.html#converting-to-strings):
```ipython
In [1]: from astropy import units as u

In [2]: u.um.to(u.nm)
Out[2]: 999.9999999999999

In [3]: f"{u.um.to(u.nm):.0f}"
Out[3]: '1000'
```
If using floating point numbers really is unacceptable then it is possible to use the [Python standard library `Decimal` class](https://docs.python.org/3/library/decimal.html) instead, but I doubt that allowing a `Quantity` to have a `Decimal` value would be worth the effort.

**Comment by WilliamJamieson** [comment]
I believe you can technically specify the dtype of a quantity when making it, so you can choose one of the integer dtypes. I don't know if the conversions respect this.

**Comment by hugobuddel** [comment]
In our case, all the quantities are floats with integer values. They are compared using <= or >=, and the result of the comparison depended on whether a unit conversion resulted in a scale factor of 999.999[..] or 1000.00[..]01. It is bad form to compare floats with equality. But in our case it could have worked, because the fractional part should have been 0.

But yeah, our code should be made more robust. Such physically insignificant changes should indeed have a negligible effect. It actually has some optional features that would taper over this problem, but I was hesitant to enable that because I didn't understand what was going on.

FWIW, specifying the dtype doesn't help: `u.Quantity(1, u.um, dtype=int).to(u.nm).value` is also 999.9999999999999.

**Comment by eerovaher** [comment]
It sounds like the simplest thing for you to do might be to replace comparisons like `x >= 1000` with `x >= 999.5`.

**Comment by pllim** [comment]
Or maybe `allclose(x, 1000) or x > 1000` ?

**Comment by hugobuddel** [comment]
> It sounds like the simplest thing for you to do might be to replace comparisons like `x >= 1000` with `x >= 999.5`.

The reality is a bit more complicated, there is no direct comparison with a fixed number, only between quantities. I tried to minimize my reproducible example.

Both numbers in the comparison start out as a copy of the same number from a configuration file, representing a wavelength in `u.um`. One copy is converted to `u.nm` (in order to talk to a third party library) and then back to `u.um` (by that library) and then they are compared (by yet another library). The actual comparison is done to determine how two ranges of wavelengths overlap. By design, the ranges should be identical, but to my surprise, they were not.

(There is always the XY problem, it is even more complicated than described above. And I suspect what this code is ultimately trying to achieve might be physically nonsense, so perhaps it shouldn't exist. But at least now the code works, so I can test my suspicion...)

By the principle of least surprise, I still think it would be better if the unit conversion is done exactly when converting between standard SI-prefixes. But there are plenty of workarounds (several in this thread), and those are essential anyway to make a program more robust. So it is fine to keep exact unit conversion at a low priority.

An idea (to make my comments a bit more constructive): A more accurate solution to unit conversion might be to (automatically?) construct a lookup table for  `UnitBase._to()` for units that can be converted exactly. That table could then also work for `u.imperial.mi.to(u.imperial.yd)`, which currently returns 1759.9999999999998.

Constructing such a unit-conversion lookup table might require changing how the SI prefixes are dealt with in `units.core`, because the factors that are currently there (e.g. `1e-9` and `1e-6`) are causing the inaccuracies. But that is literally the oldest code in the unit package, so it should not be changed on a whim. On the other hand, the prefixes code is fairly well isolated, so the change seems feasible.
